src/components/common/Modal.tsx

```diff
@@ -0,0 +1,78 @@
+import React, { useCallback, useEffect, useMemo, useState } from 'react';
+
+import clsx from 'clsx';
+import { Grow, LinearProgress, Modal as MaterialModal, Paper } from '@material-ui/core';
+
+import usePrevious from '@app/hooks/usePrevious';
+import { ModalOptions, useModal } from '@app/contexts/Modal';
+
+import useStyles from './_jss/Modal.jss';
+
+type ModalProps = Pick<ModalOptions, 'disablePaper' | 'disableIndirectDismissal'>;
+
+interface ModalViewProps {
+  isLoading: boolean;
+  modalProps: ModalProps;
+}
+
+const ModalView: React.FC<ModalViewProps> = ({ children, isLoading, modalProps }) => {
+  const { dismiss, isOpen } = useModal();
+  const [transitionIn, setTransitionIn] = useState(false);
+  const [isVisible, setIsVisible] = useState(isOpen);
+  const lastIsOpen = usePrevious(isOpen);
+  const lastChildren = usePrevious(children);
+  const classes = useStyles();
+
+  const handleClose = useCallback(() => dismiss(true), [dismiss]);
+
+  const isOpening = isOpen && !lastIsOpen;
+  const isClosing = !isOpen && lastIsOpen;
+
+  useEffect(() => {
+    if (isOpening) {
+      setIsVisible(true);
+      setTransitionIn(true);
+    } else if (isClosing) {
+      setTransitionIn(false);
+      setTimeout(() => setIsVisible(false), 400);
+    }
+  }, [isOpening, isClosing]);
+
+  const content = isClosing ? lastChildren : children;
```

Comment:
If I'm reading this correctly, this will keep the dismissed content rendered in the modal even after it's closed. When you pass in `isOpen = false` `isClosing` will be true and then it won't be false until `isOpen` is true again.

Reply:
I try to work around this by calling `setIsVisible(false)` within the useEffect above - do you think that is enough to get around this?

Reply:
In general though I know this component isn't as clear as it should be, definitely looking for ways to scoop out anything unnecessary.

Reply:
Couldn't you just null out the content 400ms after you trigger the transition? You don't need the visible state, or really the transition in state for that matter. You just need a content state really.

Reply:
I did end up removing the transition in state, however the current implementation does depend on differentiating isOpen and isVisible (or something similar). This is to distinguish the "in" prop on the Material `<Grow in={...} ... /> component from the "open" prop on the `<MaterialModal open={...} ... />` component. Take a look at how our Modal component file is structured now, I think you'll find it slightly less bloated but I'm still looking for ways to improve it.

To your point about nulling out the content, this state is managed in the context right now and the component consumes it as the children prop and just holds on to the previous value, which I think may be the more lightweight and appropriate solution. What do you think?

Reply:
The modal component already has a mechanic for waiting on a transition before closing with the `closeAfterTransition` props. All the hijinks with holding onto children for one extra render for a certain case is more convoluted than it needs to be. All you really need to do is accept content state from the parent while it's open and null it out once you're done with your transition. Here's what I mean: https://gist.github.com/juzerzarif/a9eaef306979ba4016b1168c83e1997d

Reply:
ahhh I should have just looked at the Grow API, that `onExited` props is exactly what I needed - nice find. Check out https://github.com/resident51/resident51/pull/4/commits/b6054496562b79fcb9c1e9a202ff4e7d7bf1fbde

A couple notes:
- Added the `disableRestoreFocus` prop because by default the modal returns focus to the triggering element, in this case leaving the ripple effect on the Login button
- could you use a branch instead of a gist next time? it makes viewing changes easier, plus I can checkout the branch to view the change and cherry-pick